### PR TITLE
Simple training loop

### PR DESCRIPTION
**Pull request by ayulockin**

### CODE REVIEW

**Comment by ariG23498** [comment]
Let me know if you need a review on this one.

**Comment by ayulockin** [comment]
Yes you can start reviewing especially the model file.

File: src/llm/wmt_model.py [hunk 1]
```diff
@@ -382,20 +443,65 @@ def forward(
         self,
         encoder_input: Float[Tensor, "batch enc_seq_len"],  # type: ignore
         decoder_input: Float[Tensor, "batch dec_seq_len"],  # type: ignore
+        pad_id: int = 1,
     ) -> Float[Tensor, "batch dec_seq_len vocab_tgt_size"]:  # type: ignore
+        ############# ENCODER #############
+
+        # calculate the attention mask for the encoder and decoder
+        encoder_attention_mask = encoder_input != pad_id
```
Comment (ariG23498) [inline src/llm/wmt_model.py:451]: I think we should delegate this task to data loading pipeline.

From the `transformers` perspective, we have a `tokenizer` which transforms raw text into `intputs`. `inputs` here is a culmination of `input_ids` and `attention_masks`.

WDYT?

**Review by ariG23498 — CHANGES_REQUESTED** [review]
Nits and queries.

File: src/llm/wmt_data_utils.py [hunk 2]
```diff
@@ -147,7 +139,16 @@ def _collate_fn(batch, source, target, tokenizers):
     tokenized_source_batch = torch.tensor(tokenized_source_batch)
     tokenized_target_batch = torch.tensor(tokenized_target_batch)
 
-    return tokenized_source_batch, tokenized_target_batch
+    # create attention masks
+    encoder_attention_mask = tokenized_source_batch != 1
+    decoder_attention_mask = tokenized_target_batch != 1
```
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_data_utils.py:144]: yes `pad_tokens`. it should be parameterised. I will add a TODO for now.

[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_data_utils.py:144]: actually my bad, the mask calculation is happening in the model itself. so this is not needed here. I will remove it.

File: src/llm/wmt_data_utils.py [hunk 3]
```diff
@@ -147,7 +139,16 @@ def _collate_fn(batch, source, target, tokenizers):
...
+
+    return (
+        tokenized_source_batch,
+        tokenized_target_batch,
+        encoder_attention_mask,
+        decoder_attention_mask,
+    )
```
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_data_utils.py]: now just doing 

return (
        tokenized_source_batch,
        tokenized_target_batch,
)

Would you still prefer it to return dict?

File: src/llm/wmt_model.py [hunk 4]
```diff
@@ -51,6 +51,7 @@ def forward(
         inputs: Float[Tensor, "batch seq_len model_dim"],  # type: ignore
     ) -> Float[Tensor, "batch seq_len model_dim"]:  # type: ignore
         x = self.layer1(inputs)
+        # TODO: might add dropout here
```
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_model.py:54]: No, I didn't see it but makes sense right? The feed forward is the bulkiest part so dropout there makes sense.

File: src/llm/wmt_model.py [hunk 5]
```diff
@@ -88,10 +90,26 @@ def forward(
...
+                value=-torch.inf,
+            )
+
         attention_scores = torch.softmax(
             scaled_similarity, dim=-1
         )  # Careful with the `dim`
 
+        # If a row is fully -inf, the softmax is all NaN. We can clamp that row to 0.
+        # This usually happens with short sequences. We can safely ignore these rows by
+        # clamping them to 0.
```
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_model.py:108]: as per o1, it is safe to handle it like this. we are anyway removing the padded index while doing loss calculation so the gradient will not flow from it.

File: src/llm/wmt_model.py (hunk 1 above)
Comment (ayulockin) [reply src/llm/wmt_model.py:451]: yes you are right. that would be more efficient. I did it here in this script because I needed to debug it properly. We will move it to the collate_fn.

File: src/llm/wmt_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ariG23498) [reply src/llm/wmt_model.py:54]: Yes. Can we move this to an issue rather than a TODO? Like #7 ?

File: src/llm/wmt_model.py [hunk 6]
```diff
@@ -88,10 +90,26 @@ def forward(
...
                 value=-torch.inf,
             )
+
+        if attention_mask is not None:
+            scaled_similarity = scaled_similarity.masked_fill(
+                # reversing the mask because we want to mask out the padding tokens (True in attention mask
+                # means padding token)
+                mask=~attention_mask,
+                value=-torch.inf,
+            )
```
[the remark this answers is not in the document]
Comment (ariG23498) [reply src/llm/wmt_model.py:100]: Got it! Thanks for the clarification. I think we can have a comment here on top saying why we need `attention_mask` for and why it is important.

WDYT?

File: src/llm/wmt_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ariG23498) [reply src/llm/wmt_model.py:108]: As per my intuition, I would not want to waste computation. If we are not backpropagating due to a data point, I should not forward propagate it in the first place.

File: src/llm/wmt_model.py (hunk 1 above)
Comment (ariG23498) [reply src/llm/wmt_model.py:451]: Cool -- after you debug this bit, let's move this to the data pipeline.

File: src/llm/wmt_data_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ariG23498) [reply src/llm/wmt_data_utils.py]: Ah! Glad that I asked this 😃

File: src/llm/wmt_data_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ariG23498) [reply src/llm/wmt_data_utils.py]: Upto you. No opinions.

File: src/llm/wmt_model.py (hunk 1 above)
Comment (ayulockin) [reply src/llm/wmt_model.py]: Moved it to the data pipeline.

File: src/llm/wmt_data_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_data_utils.py]: returning dicts. check out the output of the _collate_fn

File: src/llm/wmt_data_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_data_utils.py]: moved to the data pipeline.

File: src/llm/wmt_model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_model.py:54]: #10

File: src/llm/wmt_model.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ayulockin) [reply src/llm/wmt_model.py]: resolving this since we fixed the padding issue.

**Comment by ariG23498** [comment]
@ayulockin do you want me to review the PR?

**Comment by ayulockin** [comment]
Yes please @ariG23498

File: src/llm/wmt_data_utils.py [hunk 7]
```diff
@@ -103,59 +97,85 @@ def decode_wmt_tokens(
...
+    decoder_self_attention_mask = decoder_attention_mask.unsqueeze(1).expand(
+        B, dec_seq_len, dec_seq_len
+    )
+
+    # Decoder cross attention mask
+    decoder_attention_mask = decoder_attention_mask.unsqueeze(2)  # [batch, dec_len, 1]
+    encoder_attention_mask = encoder_attention_mask.unsqueeze(1)  # [batch, 1, enc_len]
+    decoder_cross_attention_mask = (
+        decoder_attention_mask & encoder_attention_mask
+    )
```
Comment (ariG23498) [inline src/llm/wmt_data_utils.py:156]: I think we can just exploit tokenizer's `.mask` api here? WDYT?

File: src/llm/wmt_model.py [hunk 8]
```diff
@@ -51,89 +50,107 @@ def forward(
         inputs: Float[Tensor, "batch seq_len model_dim"],  # type: ignore
     ) -> Float[Tensor, "batch seq_len model_dim"]:  # type: ignore
         x = self.layer1(inputs)
+        # TODO: might add dropout here
```
Comment (ariG23498) [inline src/llm/wmt_model.py:53]: Let's remove this in favour of #10

**Review by ariG23498 — CHANGES_REQUESTED** [review]
Need to look into the model a little more. Will do that after this review is satisfied.

File: src/llm/wmt_data_utils.py (hunk 7 above)
Comment (ayulockin) [reply src/llm/wmt_data_utils.py:156]: can you share documentation or working code?

Comment (ayulockin) [reply src/llm/wmt_data_utils.py:156]: working on this.

Comment (ariG23498) [reply src/llm/wmt_data_utils.py:156]: https://colab.research.google.com/drive/17m1ZwTZrmql1LHDtRA5BRWSXgPccSw5U?usp=sharing

**Comment by ayulockin** [comment]
Hey @ariG23498 this PR is ready from my end. Any more addition to this will bloat it up.

- I think we have fixed all the model related issues. The weird NaN thing that we faced is not resolved accurately. 
- The mask generation in the data pipeline is simplified.
- The mask propagation through the abstractions is simplified. Plus the attention block better handles the masking situations:
    - encoder self attention pad masking
    - decoder self attention pad masking + causal masking
    - decoder cross attention pad masking
- The model trains well (i.e, the loss goes down and doesn't NaN out after full epoch or weird spikes or gradient issues).
- The overfitting on 1 batch is successful as well -- goes to 0 loss.
- We do have code for mixed precision but it's not reliable yet but let's keep it because turning if off is as simple as `use_mixed_precision=False` in the trainer config.
- I have not added LR scheduling simply because LR is dependent on batch size and as long as we ain't training on 8 P100, it is pointless to use exactly the same hyperparameters. We need to come up with our own calculation. Here's a rough calculation documenting few things though:

## Number of train steps calculation

Original (paper) training regimen for the base model:
- 100k steps
- 25k source + 25k target tokens = 50k tokens per step
- Total tokens processed: 100k × 50k = 5×10^9

I wrote some code to check how many tokens we have in the src and target language. Note that the batch size is 32 here which I was able to fit (without much optimisation) on a A10.

```
total_source_tokens = 0
total_target_tokens = 0

source_input_tokens_list = []
target_input_tokens_list = []

for idx, batch in tqdm(enumerate(dataloaders["train"])):
    source_input = batch["source_input"]["input_ids"]
    target_input = batch["target_input"]["input_ids"]
    
    source_input_tokens = (source_input > 1).sum().item()
    target_input_tokens = (target_input > 1).sum().item()

    total_source_tokens += source_input_tokens
    total_target_tokens += target_input_tokens

    source_input_tokens_list.append(source_input_tokens)
    target_input_tokens_list.append(target_input_tokens)


import numpy as np
avg_source_tokens = np.mean(source_input_tokens_list)
print(f"Average source tokens per batch: {avg_source_tokens:.2f}")
avg_target_tokens = np.mean(target_input_tokens_list)
print(f"Average target tokens per batch: {avg_target_tokens:.2f}")
total_source_tokens, total_target_tokens
```

Entire training set
- Total source tokens: ~135.6M (135628910 exactly)
- Total target tokens: ~141.1M (141145731 exactly)
- Total tokens: ~276.8M (276774641 exactly)

source is "de" and target is "en". I am also only counting the valid tokens and not special tokens in each batch to get these values.

Per-batch averages
- Average source tokens per batch: ~962.6
- Average target tokens per batch: ~1001.7
- Average total tokens per batch: ~1,964

So in our current setup we have ~2000 tokens in the batch (both src and target) which is way off from the 50k in the original training regime. Thus we need these many steps to hit the same tokens processed:

image

yes 2.5 million steps!

So how many epochs (adhering to our current setup)?

Each epoch expose our model to 2.77×10^8 tokens (ignoring the special tokens) and which account to ~141k steps.

image

so if we want to process 5×10^9 tokens we need 18 epochs:

image

In our current setup 1 epoch takes 5hr 48 mins:

image

so 18 x 5 hr 48 mins = 18 x 348 mins = 6264 mins / 60 = 104.4 hrs / 24 = 4.35 days. 

The original training regime takes 3.5 days. 

Idea: bucketed batching where we ensure a fixed token is processed each step instead of fixed number of examples.

**Review by ariG23498 — APPROVED** [review]
LGTM!
